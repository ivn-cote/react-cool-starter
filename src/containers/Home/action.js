/* @flow */
import requestAPI from '../../communication';

import type {
  Dispatch,
  GetState,
  ThunkAction,
  Reducer,
} from '../../types';

export const USERS_REQUESTING = 'USERS_REQUESTING';
export const USERS_FAILURE = 'USERS_FAILURE';
export const USERS_SUCCESS = 'USERS_SUCCESS';

export const API_URL = 'users/';

// Export this for unit testing more easily
export const fetchUsers = (endpointPath: string = API_URL): ThunkAction =>
  (dispatch: Dispatch, getStore) => {
    dispatch({ type: USERS_REQUESTING });

    return requestAPI(endpointPath, getStore().config)
      .then((res) => {
        dispatch({ type: USERS_SUCCESS, data: res.data });
      })
      .catch((err) => {
        dispatch({ type: USERS_FAILURE, err });
      });
  };

// Preventing dobule fetching data
/* istanbul ignore next */
const shouldFetchUsers = (state: Reducer): boolean => {
  // In development, we will allow action dispatching
  // or your reducer hot reloading won't updated on the view
  if (__DEV__) return true;

  const home = state.home;

  if (home.readyStatus === USERS_SUCCESS) return false; // Preventing double fetching data

  return true;
};

/* istanbul ignore next */
export const fetchUsersIfNeeded = (): ThunkAction =>
  (dispatch: Dispatch, getState: GetState) => {
    /* istanbul ignore next */
    if (shouldFetchUsers(getState())) {
      /* istanbul ignore next */
      return dispatch(fetchUsers());
    }

    /* istanbul ignore next */
    return null;
  };
