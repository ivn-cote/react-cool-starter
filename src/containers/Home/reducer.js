/* @flow */

import _ from 'lodash';

import {
  USERS_REQUESTING,
  USERS_FAILURE,
  USERS_SUCCESS,
} from './action';
import type { Home, Action } from '../../types';

type State = Home;

const initialState = {
  loading: false,
  err: null,
  list: [],
};

export default (state: State = initialState, action: Action): State => {
  switch (action.type) {
    case USERS_REQUESTING:
      return _.assign({}, state, { loading: true });
    case USERS_FAILURE:
      return _.assign({}, state, {
        loading: false,
        err: action.err,
      });
    case USERS_SUCCESS:
      return _.assign({}, state, {
        loading: false,
        list: action.data,
        err: null,
      });
    default:
      return state;
  }
};
