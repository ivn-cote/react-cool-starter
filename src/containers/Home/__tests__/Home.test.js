import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { StaticRouter } from 'react-router-dom';

import { Home } from '../Home';

jest.mock('../../../components/UserList', () => 'UserList');

describe('<Home />', () => {
  const tree = (props, actions) => renderer.create(
    <StaticRouter location={''} context={{}}>
      <Home {...props} {...actions} />
    </StaticRouter>,
  ).toJSON();

  test('should call fetchUsersIfNeeded when componentDidMount', () => {
    const mockAction = jest.fn();
    const props = {
      home: {},
    };
    const actions = {
      fetchUsersIfNeeded: mockAction,
    };

    mount(
      <StaticRouter location={''} context={{}}>
        <Home {...props} {...actions} />
      </StaticRouter>,
    );

    expect(mockAction).toHaveBeenCalled();
  });

  test('renders nothing if no data', () => {
    const props = {
      home: { },
    };
    const actions = { fetchUsersIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });

  test('renders the loading status if requesting data', () => {
    const props = {
      home: { loading: true },
    };
    const actions = { fetchUsersIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });

  test('renders an error if loading failed', () => {
    const props = {
      home: { err: 'some error' },
    };
    const actions = { fetchUsersIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });

  test('renders the <UserList /> if loading was successful', () => {
    const props = {
      home: {
        list: [{ id: '1', name: 'Welly' }],
      },
    };
    const actions = { fetchUsersIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });
});
