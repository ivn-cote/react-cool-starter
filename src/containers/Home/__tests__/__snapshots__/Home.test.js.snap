// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> renders an error if loading failed 1`] = `
<div
  className={undefined}
>
  <p>
    Oops, Failed to load list!
  </p>
</div>
`;

exports[`<Home /> renders the <UserList /> if loading was successful 1`] = `
<div
  className={undefined}
>
  <UserList
    list={
      Array [
        Object {
          "id": "1",
          "name": "Welly",
        },
      ]
    }
  />
</div>
`;

exports[`<Home /> renders the loading status if data invalid 1`] = `
<div
  className={undefined}
>
  <p>
    Loading...
  </p>
</div>
`;

exports[`<Home /> renders the loading status if requesting data 1`] = `
<div
  className={undefined}
>
  <p>
    Loading...
  </p>
</div>
`;
